packages/features/emailValidation/lib/service/EmailValidationCachingProxy.ts

```diff
@@ -0,0 +1,105 @@
+import type { IRedisService } from "@calcom/features/redis/IRedisService";
+import logger from "@calcom/lib/logger";
+import { safeStringify } from "@calcom/lib/safeStringify";
+
+import type { EmailValidationRequest, EmailValidationResult, EmailValidationStatus } from "../dto/types";
+import type {
+  IEmailValidationCachingProxy,
+  IEmailValidationService,
+} from "./IEmailValidationService.interface";
+
+export const REDIS_EMAIL_VALIDATION_CACHE_KEY = (email: string) => `email_validation:${email.toLowerCase()}`;
+
+export const EMAIL_VALIDATION_CACHE_TTL_MS = 7_200_000; // 2 hours
+
+function getPiiFreeEmail(email: string): string {
+  return email.toLowerCase().slice(0, 10);
+}
+
+export class EmailValidationCachingProxy implements IEmailValidationCachingProxy {
+  private readonly logger = logger.getSubLogger({ prefix: ["EmailValidationCachingProxy"] });
+
+  constructor(
+    private readonly deps: {
+      emailValidationService: IEmailValidationService;
+      redisService: IRedisService;
+    }
+  ) {}
+
+  /**
+   * Returns an async generator that yields a null value if the email validation result is not cached.
+   * If the email validation result is cached, it yields the cached result.
+   * If the email validation result is not cached, it calls the email validation service and caches the result.
+   *
+   * @param request - The email validation request
+   * @returns
+   */
+  async *validateEmailGenerator(
+    request: EmailValidationRequest
+  ): AsyncGenerator<null, EmailValidationResult, unknown> {
+    const cacheKey = REDIS_EMAIL_VALIDATION_CACHE_KEY(request.email);
+    try {
+      const cachedResult = await this.deps.redisService.get<EmailValidationResult>(cacheKey);
+      if (cachedResult) {
+        this.logger.info(
+          "Email validation cache hit",
+          safeStringify({
+            email: getPiiFreeEmail(request.email),
+            status: cachedResult.status,
+          })
+        );
+        return cachedResult;
+      }
+
+      this.logger.info(
+        "Email validation cache miss, calling service",
+        safeStringify({
+          email: getPiiFreeEmail(request.email),
+        })
+      );
+
+      // We yield so that we communicate that we don't have cached result and we further need to call next() to get the result from the actual service
+      yield null;
```

Comment:
Yielding here ensures that when someone awaits the first next() call they get the cached response.
and then they can call the next next() without await